text/0374-jsii-ts-version.md

```diff
@@ -0,0 +1,324 @@
+(in particular, the `.d.ts` declarations files and the `.jsii` assemblies)
+remain compatible with previous versions, so that consumers of libraries need
+not update at the same time as their dependencies.
+
+### What happens to `jsii` v1?
+
+Version `1.0.0` of the `jsii` compiler was released over two years ago (in
+February 2020), and will transition to the *Maintenance* tier of the
+[AWS SDKs and Tools maintenance policy][aws-policy] 6 months after this release.
+
+During this initial 6 months period, `jsii` will continue to receive full
+support, including bug fixes and feature support, as it has in the past two
+years.
+
+Once it enters the *Maintenance* tier, it will only receive fixes for critical
+bugs and security issues. The `v1` release line will remain in *Maintenace* for
+12 months, after which it will transition into the *End-of-Life* tier.
+
+### What is the support policy for these new releases?
+
+Starting with release `4.7.0`, new features will only be added to the *current*
+release line (the *latest* `major.minor` stream), which corresponds to the
+currently active [TypeScript] release line.
+
+Critical bug fixes and security patches will be provided for previous release
+lines for 12 months, which covers 3 or 4 previous lines. Features will not be
+back-ported to non-current release lines.
+
+Previous release lines will be considered *end-of-life* 12 months after they
+were superceded by a new *current* line.
+
+## Internal FAQ
+
+### How will we maintain backwards compatibility?
+
+The change only affects the [TypeScript] compiler version used internally by
+`jsii` and the versioning scheme for the `jsii` package itself.
+
+The compiler will continue to emit `.jsii` assemblies that conform to the schema
+defined in the `@jsii/spec` package, which will hence continue to be compatible
+with all other tools part of the jsii toolchain (including `jsii-pacmak`, ...).
+
+In order to maximize compatibility between [TypeScript] compiler versions, and
+since [TypeScript] occasionally introduces backwards-incompatible syntax changes
+(additions, modifications) to the declarations files (`.d.ts`), the `jsii`
+compiler will proactively produce down-leveled declarations files targeting
+[TypeScript] compiler versions used by previous (not yet *end-of-life*) releases
+of `jsii`. This can be achieved using the [`downlevel-dts`][downlevel-dts]
+utility.
+
+### Why are we doing this?
+
+Before this change, `jsii` users had been stuck with [TypeScript] `3.9` for a
+very long time, due to the introduction of several breaking changes in the
+language specification: upgrading the [TypeScript] compiler that `jsii` builds
+on would cause existing code to break.
+
+This could not be addressed by issuing a new major release of `jsii`, as this
+would cause the ecosystem to be fragmented between packages that have migrated
+to the new major release, and those that haven't. Additionally, a package could
```

Comment:
Why is this? That would only be true if the runtime changed versions as well, right? Why don't we decouple the runtime version from the compiler version?

Reply:
And in fact, this proposal doesn't talk about it but it sure seems like an implication of this is that we should be decoupling the runtime version from the jsii compiler version even starting with this proposal -- otherwise jsii 3.9.x packages require runtime version 3.9.x and jsii 4.7.x reuqires runtime 4.7.x and we would be inducing the same fragmentation this paragraph warns about.

Reply:
Basically this proposal is entirely about decoupling the compiler version from the rest of the toolchain's versions...

In the current state of things, we MUST either major-bump nothing, or major-bump everything... and cause a schism.

---

The previous version of this RFC was going the extreme direction of suggestion we decouple the versions of **everything** and break the mono-repo out to individual repositories (possibly in a new GitHub organization, but that is beyond the point).

The same reasoning line was exposed there but literally every piece of feedback I received on that initial version included comments along the lines of "but why?", and I honestly don't have more things to say about "why" than this (and that feedback seemed to indicate this is not a sufficient justification?)

As a result, I decided to narrow the scope down to ONLY address the one problem that is currently a known problem: we've been stuck on TypeScript 3.9 for **way** too long, and need a clean way out.